src/transaction/atomic_replication_helper.cpp

```diff
@@ -301,160 +354,465 @@ namespace cublog
+    //  start        start           |              start          |   |
+    //                              end                          end   end
+    //
+    //  Legend:
+    //    - C1, C2, .. : control log records
+    //    - W1, W2, .. : work/payload log records
+    //
+    //  The above sequence would be processed in the following way:
+    //  - C1 added
+    //      - can_purge is called, nothing to do
+    //  - W1-W3 added and accumulated
+    //  - C2 added
+    //      - decide that W1-W3 can be processed - process and remove
+    //      - can_purge is called, nothing to do
+    //  - W4-W6 added and accumulated
+    //  - C3 added
+    //      - decide that W4-W6 can be processed - process and remove
+    //      - can_purge is called, decide that C2, C3 can be removed
+    //  - W7-W10 added and accumulated
+    //  - C4 added
+    //      - decide that W7-W10 can be processed - process and remove
+    //      - can_purge is called, nothing to do as C4 does not have
+    //        a matching end pair "control" log record
+    //  - W11, W12 added and accumulated
+    //  - C5 added
+    //      - decide that W11, W12 can be processed - process and remove
+    //      - can_purge is called, decide that C4, C5 can be removed
+    //  - C6 added
+    //      - no actual paylod log records to apply
+    //      - can_purge is called, decide that C1, C6 can be removed
+    //
+    //  after this, the entire sequence is removed as it is left empty;
+    //
+    // - if, at some point within the same transaction, a new atomic sequence
+    //  control log records is encountered, a new, separate atomic sequence
+    //  object is created and handled
+    //
+    // (*) controlled manner = in such a way as to maintain consistency; this is mostly
+    // the result of empirical observations of encountered "control" log sequences;
+    // these sequences are laid out in the atomic_replication_helper class comment and
+    // their logic is implemented and acted upon in this function and the function
+    // calling this one
+    while (true)
       {
-	assert (m_is_sysop);
-	assert (m_postpone_started);
-	return true;
+	if (m_log_vec.empty ())
+	  {
+	    return true;
+	  }
+
+	const size_t initial_log_vec_size = m_log_vec.size ();
+
+	// nothing can be inferred from only one entry alone
+	if (initial_log_vec_size == 1)
+	  {
+	    return false;
+	  }
+
+	if (prm_get_bool_value (PRM_ID_ER_LOG_PTS_ATOMIC_REPL_DEBUG))
+	  {
+	    dump ("sequence::can_purge - START");
+	  }
+
+	const atomic_log_entry_vector_type::const_iterator last_entry_it = m_log_vec.cend () - 1;
+	const atomic_log_entry &last_entry = *last_entry_it;
+
+	// atomic replication sequence with an already executed postpone sequence that (maybe) contained
+	// - itself - other atomic replication sequences)
+	if (LOG_SYSOP_END == last_entry.m_rectype
+	    && LOG_SYSOP_END_COMMIT == last_entry.m_sysop_end_type)
```

Comment:
```suggestion
	if (last_entry.m_rectype == LOG_SYSOP_END
	    && last_entry.m_sysop_end_type == LOG_SYSOP_END_COMMIT)
```

Reply:
Typical "C/C++-ism" to prevent mistakes of the type
```
if (last_entry.m_rectype = LOG_SYSOP_END) ...
``
To prevent the compiler's automatic conversion/promotion to boolean true.